go/private/context.bzl

```diff
@@ -439,6 +439,14 @@ def go_context(
         #
         # See https://go.dev/doc/toolchain for more info.
         "GOTOOLCHAIN": "local",
+
+        # NOTE(#4049): Since Go 1.23.0, os.Readlink (and consequently
+        # filepath.EvalSymlinks) stopped treating Windows mount points and
+        # reparse points (junctions) as symbolic links. Bazel uses junctions
+        # when constructing exec roots, and we use filepath.EvalSymlinks in
+        # GoStdlib, so this broke us. Setting GODEBUG=winsymlink=0 restores
+        # the old behavior.
+        "GODEBUG": "winsymlink=0",
```

Comment:
Wouldn't this also apply to compilation of user code? If we only need it for the builder binary, maybe we could move it closer to that.

Reply:
Do you have an idea on how this should work?

I'm actually not familiar with the `GODEBUG` mechanism. I thought it mainly applied at run-time, so setting it here would only affect rules_go actions. But [Go, Backwards Compatibility, and GODEBUG](https://go.dev/doc/godebug) says this:

> When a GODEBUG setting is not listed in the environment variable, its value is derived from three sources: the defaults for the Go toolchain used to build the program, amended to match the Go version listed in go.mod, and then overridden by explicit //go:debug lines in the program.

This is a bit vague: I'm not sure if "the Go toolchain" here means the compiler, linker, std, or `go` tool. `go.mod` doesn't apply here. I'm also not sure if the `//go:debug` lines are interpreted by the `go` tool or compiler.

I'm guessing we should set `GODEBUG=winsymlink=0` in the scripts run by `go_tool_binary`, since that builds the builder binary and the `std` packages it depends on, especially `runtime`, which seems to be responsible for parsing `GODEBUG`. If I understand correctly, that would have no side effect on the `GoStdlib` action (other than working around the bug) or on compiled user code. Does that sound workable?

Reply:
Yes, that's my understanding as well: we should only set the variable when building the builder (in `go_tool_binary`).

Reply:
I think this works: I've added a `//go:debug winsymlink=0` directive in builder.go. These directives can be added to `main` packages and control the default for that binary.

Setting `GODEBUG=winsymlink=0` in the `go_tool_binary` scripts didn't seem to work: that only seems to apply at run-time, not compile-time.

Reply:
Nope, does not work. Versions of Go before 1.23.0 error on `//go:debug winsymlink=0` because of the unknown build constraint.

Reply:
I suspect the *only* way to make this work is to set `GODEBUG` in the environment when running the builder, at least for the `GoStdLib` action. As I understand, setting `GODEBUG` in the environment affects only run-time behavior, not behavior of compiled code. WDYT?

Reply:
You are right, I misunderstood how `GODEBUG` works. This sounds right, sorry!

Reply:
All right, switched back to setting `GODEBUG` for all actions. I think `GoStdlib` and `GoStdlibList` are the only ones that need it, but if we kept it so narrow, someone would get a very confusing error message if they tried to call `replicate` anywhere else.